### PR TITLE
Karel/assets more tests

### PR DESCRIPTION
**Pull request by ysfkel**

### CODE REVIEW

File: frame/assets/src/tests/traits.rs [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::free_balance(asset_id, &account_1), first + second + third);
+
+				Ok(())
+			}).unwrap();
+		}
+	}
+}
+
+/*
+type PositiveImbalance: Imbalance<Self::Balance, Opposite = Self::NegativeImbalance>
```
Comment (dzmitry-lahoda) [inline frame/assets/src/tests/traits.rs:656]: it is like if this comment should be moved closer to place where we should implement this? or it is TODO for tests? not clear

File: frame/assets/src/mocks.rs [hunk 2]
```diff
@@ -12,13 +15,29 @@ use sp_runtime::{
 };
 use system::EnsureRoot;
 
+//
```
Comment (dzmitry-lahoda) [inline frame/assets/src/mocks.rs:18]: empty comments

File: frame/assets/src/mocks.rs [hunk 3]
```diff
@@ -12,13 +15,29 @@ use sp_runtime::{
...
 pub type Block = frame_system::mocking::MockBlock<Test>;
 pub type AccountId = u64;
 pub type AssetId = u64;
 pub type Amount = i128;
 pub type Balance = u64;
 
+pub const ALICE: AccountId = 1;
+pub const BOB: AccountId = 2;
+
+pub const EVE: AccountId = 3;
```
Comment (dzmitry-lahoda) [inline frame/assets/src/mocks.rs:32]: usually this is C(Charlie, ABC, ABCDE - is 5)

File: frame/assets/src/mocks.rs [hunk 4]
```diff
@@ -129,12 +148,26 @@ impl pallet_balances::Config for Test {
 	type ReserveIdentifier = [u8; 8];
 }
 
-// Build genesis storage according to the mock runtime.
+pub const BALANCES: [(AccountId, Balance); 4] = [(0, 1000), (1, 1000), (2, 1000), (3, 1000)];
```
Comment (dzmitry-lahoda) [inline frame/assets/src/mocks.rs:151]: 0? who is it? Alice = 0, Bob =1 , etc. May be used ABCD here too?

File: frame/assets/Cargo.toml [hunk 5]
```diff
@@ -36,6 +36,8 @@ version = "2.0.0"
 pallet-balances = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.13" }
 orml-tokens = { git = "https://github.com/open-web3-stack/open-runtime-module-library", rev = "17a791edf431d7d7aee1ea3dfaeeb7bc21944301" }
 governance-registry = { package = "pallet-governance-registry", path = "../governance-registry", default-features = false }
+proptest = "0.9.6"
+composable-tests-helpers = { version = "0.0.1", path = "../composable-tests-helpers", default-features = false }
```
Comment (dzmitry-lahoda) [inline frame/assets/Cargo.toml:40]: remove version = "0.0.1"

File: frame/assets/src/tests/traits.rs [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AssetId>>::free_balance(asset_id, &account_1), free_balance + (second - remaining));
+
+				free_balance = <Pallet::<Test> as MultiCurrency<AccountId>>::free_balance(asset_id, &account_1);
+				remaining = <Pallet::<Test> as MultiReservableCurrency<AccountId>>::unreserve(asset_id, &account_1, first);
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::free_balance(asset_id, &account_1), free_balance + (first - remaining));
+				prop_assert_eq!(<Pallet::<Test> as MultiReservableCurrency<AccountId>>::reserved_balance(asset_id, &account_1), 0);
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::free_balance(asset_id, &account_1), first + second + third);
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:650]: You should return to the error to proptest here, using `?;` and not unwrap...

File: frame/assets/src/tests/traits.rs [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+			   prop_assert_ok!(<Pallet::<Test> as MultiReservableCurrency<AssetId>>::reserve(asset_id, &account_1, first + second ));
+			   //repatriate to free balance
+			   let repatriate_free = <Pallet::<Test> as MultiReservableCurrency<AssetId>>::repatriate_reserved(asset_id, &account_1, &account_2, second, BalanceStatus::Free).unwrap();
+			   prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AssetId>>::free_balance(asset_id, &account_2), first + (second - repatriate_free));
+
+			   prop_assert_eq!(<Pallet::<Test> as MultiReservableCurrency<AssetId>>::reserved_balance(asset_id, &account_1), first + repatriate_free );
+
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:619]: You should return to the error to proptest here, using `?;` and not unwrap...

File: frame/assets/src/tests/traits.rs [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+					balance
+				} else {
+					prop_assert_eq!(difference, third - (first + second ));
+					prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AssetId>>::total_issuance(asset_id),total_issuance-(third-difference));
+					prop_assert_eq!(<Pallet::<Test> as MultiReservableCurrency<AssetId>>::reserved_balance(asset_id, &account_1), 0);
+					0
+				};
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:597]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+				prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AssetId>>::deposit(asset_id,&account_1, third));
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AssetId>>::total_issuance(asset_id), first + second + third);
+
+				//reserve
+				prop_assert_ok!(<Pallet::<Test> as MultiReservableCurrency<AccountId>>::reserve(asset_id, &account_1, first + second));
+				prop_assert_eq!(<Pallet::<Test> as MultiReservableCurrency<AccountId>>::reserved_balance(asset_id, &account_1), first + second);
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::free_balance(asset_id, &account_1), third);
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:564]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 10]
```diff
@@ -0,0 +1,662 @@
...
+			(first, _, _) in valid_amounts_without_overflow_3()
+		) {
+			new_test_ext().execute_with(|| {
+
+				prop_assert_eq!(<Pallet::<Test> as MultiReservableCurrency<AssetId>>::can_reserve(asset_id, &account_1, first), false);
+				prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AssetId>>::deposit(asset_id,&account_1, first));
+				prop_assert_eq!(<Pallet::<Test> as MultiReservableCurrency<AssetId>>::can_reserve(asset_id, &account_1, first), true);
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:538]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 11]
```diff
@@ -0,0 +1,662 @@
...
+			 prop_assert!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, first + second + third).is_err());
+			 prop_assert!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, first ).is_err());
+			 prop_assert!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, second ).is_err());
+			 prop_assert!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, third).is_err());
+
+			prop_assert_ok!(<Pallet::<Test> as MultiLockableCurrency<AccountId>>::remove_lock(*b"prelocks", asset_id, &account));
+			prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, first + second + third));
+
+			Ok(())
+		  }).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:511]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+
+			 <Pallet::<Test> as MultiLockableCurrency<AccountId>>::extend_lock(*b"prelocks", asset_id, &account, second);
+			 prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, third ));
+			 prop_assert!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, first + second + third).is_err());
+
+			 <Pallet::<Test> as MultiLockableCurrency<AccountId>>::extend_lock(*b"prelocks", asset_id, &account,  third);
+			 prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, first + second));
+			 prop_assert!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, first + second + third).is_err());
+			Ok(())
+		  }).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:487]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 13]
```diff
@@ -0,0 +1,662 @@
...
+			 prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account,  first + third ));
+			 prop_assert!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, first + second + third).is_err());
+
+			  // ensure set_lock updates lock with same id
+			 <Pallet::<Test> as MultiLockableCurrency<AccountId>>::set_lock(*b"prelocks", asset_id, &account,  third);
+			 prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account,  first + second ));
+			 prop_assert!(<Pallet::<Test> as MultiCurrency<AccountId>>::ensure_can_withdraw(asset_id, &account, first + second + third).is_err());
+
+			Ok(())
+		  }).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:461]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 14]
```diff
@@ -0,0 +1,662 @@
...
+				remaining = <Pallet::<Test> as MultiReservableCurrency<AccountId>>::unreserve(asset_id, &account, third);
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::free_balance(asset_id,&account),  free_balance + ( third - remaining));
+
+				//lock
+				free_balance = <Pallet::<Test> as MultiCurrency<AccountId>>::free_balance(asset_id,&account);
+				<Pallet::<Test> as MultiLockableCurrency<AccountId>>::set_lock(*b"prelocks", asset_id, &account, first);
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::free_balance(asset_id,&account),  first + second + third);
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:422]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 15]
```diff
@@ -0,0 +1,662 @@
...
+				prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AccountId>>::deposit(asset_id, &account_1, first));
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::total_issuance(asset_id), first);
+
+				prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AccountId>>::deposit(asset_id, &account_2, second));
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::total_issuance(asset_id), first + second);
+
+				prop_assert_ok!(<Pallet::<Test> as MultiCurrency<AccountId>>::deposit(asset_id, &account_1, third));
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::total_issuance(asset_id), first + second + third);
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:374]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 16]
```diff
@@ -0,0 +1,662 @@
...
+		#[test]
+		fn test_minimum_balance_implementation(
+			account in accounts(),
+			asset_id in asset(),
+			(_first, _second, _third) in valid_amounts_without_overflow_3()) {
+			new_test_ext_multi_currency().execute_with(|| {
+
+				prop_assert_eq!(<Pallet::<Test> as MultiCurrency<AccountId>>::minimum_balance(asset_id), 0);
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:332]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 17]
```diff
@@ -0,0 +1,662 @@
...
+				prop_assert_eq!(<Pallet::<Test> as Currency<AccountId>>::free_balance(&account_1), free_balance + (second - remaining));
+
+				free_balance = <Pallet::<Test> as Currency<AccountId>>::free_balance(&account_1);
+				remaining = <Pallet::<Test> as ReservableCurrency<AccountId>>::unreserve(&account_1, first);
+				prop_assert_eq!(<Pallet::<Test> as Currency<AccountId>>::free_balance(&account_1), free_balance + (first - remaining));
+				prop_assert_eq!(<Pallet::<Test> as ReservableCurrency<AccountId>>::reserved_balance(&account_1), 0);
+				prop_assert_eq!(<Pallet::<Test> as Currency<AccountId>>::free_balance(&account_1), first + second + third);
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:310]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 18]
```diff
@@ -0,0 +1,662 @@
...
+				//repatriate to free balance
+				let repatriate_free = <Pallet::<Test> as ReservableCurrency<AccountId>>::repatriate_reserved(&account_1, &account_2, second, BalanceStatus::Free).unwrap();
+				prop_assert_eq!(<Pallet::<Test> as Currency<AccountId>>::free_balance(&account_2),first + (second - repatriate_free));
+				//repatriate to reserved balance
+				let repatriate_reserved = <Pallet::<Test> as ReservableCurrency<AccountId>>::repatriate_reserved(&account_1, &account_2, third, BalanceStatus::Reserved).unwrap();
+				prop_assert_eq!(<Pallet::<Test> as ReservableCurrency<AccountId>>::reserved_balance(&account_2), third - repatriate_reserved);
+				prop_assert_eq!(<Pallet::<Test> as ReservableCurrency<AccountId>>::reserved_balance(&account_1), first + (repatriate_free + repatriate_reserved));
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:280]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 19]
```diff
@@ -0,0 +1,662 @@
...
+					balance
+				} else {
+					prop_assert_eq!(difference, third - (first + second ));
+					prop_assert_eq!(<Pallet::<Test> as Currency<AccountId>>::total_issuance(),total_issuance-(third-difference));
+					prop_assert_eq!(<Pallet::<Test> as ReservableCurrency<AccountId>>::reserved_balance(&account_1), 0);
+					0
+				};
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:258]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 20]
```diff
@@ -0,0 +1,662 @@
...
+				assert_issuance!(first+second);
+				//increase user balance
+				prop_assert_eq!(<Pallet::<Test> as Currency<AccountId>>::deposit_into_existing(&account_1, third).unwrap().peek(), third);
+				assert_issuance!(first+second+third);
+				//reserve
+				prop_assert_ok!(<Pallet::<Test> as ReservableCurrency<AccountId>>::reserve(&account_1, first+second));
+				prop_assert_eq!(<Pallet::<Test> as ReservableCurrency<AccountId>>::reserved_balance(&account_1), first+second);
+				prop_assert_eq!(<Pallet::<Test> as Currency<AccountId>>::free_balance(&account_1), third);
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:226]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 21]
```diff
@@ -0,0 +1,662 @@
...
+			new_test_ext().execute_with(|| {
+
+				prop_assert_eq!(<Pallet::<Test> as ReservableCurrency<AccountId>>::can_reserve(&account_1, first), false);
+				assert_issuance!(0);
+				prop_assert_eq!(<Pallet::<Test> as Currency<AccountId>>::deposit_creating(&account_1, first).peek(), first);
+				assert_issuance!(first);
+				prop_assert_eq!(<Pallet::<Test> as ReservableCurrency<AccountId>>::can_reserve(&account_1, first), true);
+
+				Ok(())
+			}).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:203]: You should return to the error to proptest here, using `?;` and not unwrap

File: frame/assets/src/tests/traits.rs [hunk 22]
```diff
@@ -0,0 +1,662 @@
...
+
+				let receiver = &(account + 1);
+				prop_assert_ok!(<Pallet::<Test> as Currency<AccountId>>::transfer(&account, receiver, third,
+				ExistenceRequirement::AllowDeath));
+				prop_assert_eq!(<Pallet::<Test> as
+				Currency<AccountId>>::total_balance(receiver), third);
+				prop_assert_eq!(<Pallet::<Test> as Currency<AccountId>>::free_balance(receiver), third);
+				Ok(())
+
+			 }).unwrap();
```
Comment (hussein-aitlahcen) [inline frame/assets/src/tests/traits.rs:164]: You should return to the error to proptest here, using `?;` and not unwrap

**Review by hussein-aitlahcen — CHANGES_REQUESTED** [review]

File: frame/assets/src/mocks.rs (hunk 3 above)
Comment (ysfkel) [reply frame/assets/src/mocks.rs:28]: could you clarify this?

Comment (ysfkel) [reply frame/assets/src/mocks.rs:28]: Clarify this please

Comment (dzmitry-lahoda) [reply frame/assets/src/mocks.rs:28]: I mean convention is used Charlie (C) before using Eve(E)

**Comment by hussein-aitlahcen** [comment]
I think you need to merge/rebase on, main

File: frame/assets/src/mocks.rs (hunk 3 above)
Comment (ysfkel) [reply frame/assets/src/mocks.rs:28]: ok

**Review by dzmitry-lahoda — DISMISSED** [review]

**Review by hussein-aitlahcen — DISMISSED** [review]

**Review by dzmitry-lahoda — DISMISSED** [review]

**Pull request merged**